pkg/recovery/recover.go

```diff
@@ -0,0 +1,332 @@
+// Package recovery provides tooling to help with control plane disaster recovery. Recover() uses a
+// Backend to extract the control plane from a store, such as etcd, and use those to write assets
+// that can be used by `bootkube start` to reboot the control plane.
+//
+// The recovery tool assumes that the component names for the control plane elements are the same as
+// what is output by `bootkube render`. The `bootkube start` command also makes this assumption.
+// It also assumes that kubeconfig on the kubelet is located at /etc/kubernetes/kubeconfig, though
+// that can be changed in the bootstrap manifests that are rendered.
+package recovery
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	policyv1beta1 "k8s.io/client-go/pkg/apis/policy/v1beta1"
+
+	"github.com/kubernetes-incubator/bootkube/pkg/asset"
+)
+
+const (
+	kubeletKubeConfigPath = "/etc/kubernetes/kubeconfig"
+)
+
+var (
+	// bootstrapComponents contains the names of the components that we will extract to construct the
+	// temporary bootstrap control plane.
+	bootstrapComponents = map[string]bool{
+		"kube-apiserver":          true,
+		"kube-controller-manager": true,
+		"kube-scheduler":          true,
+		"pod-checkpointer":        true,
```

Comment:
I don't think we need the checkpointer to re-bootstrap the control plane (and once a control-plane is running, it will get scheduled normally).